Add a check for some followed by filter

qmwrygdxpzc opened the pull request:
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust-clippy/pull/15745)*

Add a check for some followed by filter.

The program will suggest using the `then_some` method.

changelog: [`filter_some`]: https://rust-lang.github.io/rust-clippy/master/index.html#filter_some

samueltardieu reviewed (dismissed):
Thanks for your contribution. However, the lint cannot be considered for merging as-is:
- It should include autofixes suggestions when possible.
- It should cover more complex cases.

For example, with the following code

```rust
fn with_side_effect(x: u32) -> u32 {
    println!("In `with_side_effect()`");
    x * 10 + 30
}

fn f(x: u32) -> Option<u32> {
    Some(with_side_effect(x)).filter(|&x| x > 40)
}
```

the lint suggests using `then_some()` inside `f()`. However, I don't see what would be the replacement code here.

You should include tests for the most complicated situations you can think of, as the lint should not fail, don't just cover the simplest case where it is obvious it will work. For example, what is part of the expression comes from a macro? Also, `bool::then_some()` has been introduced in Rust 1.62.0, so the MSRV must be checked as well. 

Hope this review will help you go forward.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15745/af0bc983b5b0a6f3f8d05f8a0baed4e97ec51d39..d34f4eda58f60df54dc6d40e322f749472afee11)*

qmwrygdxpzc commented:
It has been about 2 weeks.  Can someone take a look?

qmwrygdxpzc commented:
I used the `reindent_multiline` function because the check that I used as an example has it.  I noticed that some checks don't use it.  When should it be used?

ada4a reviewed (changes requested):
Sorry, completely missed your comment for some reason

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15745/e299e18e546bee90c64eba55f4276cd62633b4c0..0be3e4b0642686b06320f29d95d5c9edf07ca14a)*

qmwrygdxpzc commented on clippy_lints/src/methods/filter_some.rs:
```diff
@@ -0,0 +1,61 @@
+        diag.span_suggestion(
+            expr.span,
+            "consider using `then_some` instead",
+            reindent_multiline(&sugg, true, indent_of(cx, expr.span)),
```
I added the tests.  The replacement was the same whether or not the function was used.  I removed the function call.

qmwrygdxpzc commented:
@ada4a It looks like the reviewers have not asked for changes in a long time.  What should we do as the author?

ada4a reviewed (changes requested):
Sorry it took me a while to get back to you. The code looks good to me now, thank you:) Let me start the FCP while you're dealing with the last comments.

@rustbot label -needs-fcp +lint-nominated

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15745/13d91b7ccecdd98449b7cc9b0572a59de1eeb0bd..724ed1a51d7da79295cd23ed16190dc995cf5d2e)*

qmwrygdxpzc commented on clippy_lints/src/methods/some_filter.rs:
```diff
@@ -0,0 +1,60 @@
+        "{}{condition_text}{}.then_some({value_text})",
+        if parentheses { "(" } else { "" },
+        if parentheses { ")" } else { "" },
+    );
```
I have moved it inside the function call.  But why is it better?

ada4a reviewed (changes requested):
some last comments^^

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15745/174df86e5f8643cd9634d8d93f1cafc2ca42a863..fd96e9e778ec5a5d519776bb43828c7e71d69e4c)*

Commit:
Add a check for some followed by filter

The program will suggest using the `then_some` method.

changelog: add [`some_filter`]

ada4a commented on clippy_lints/src/methods/mod.rs:
````diff
@@ -3575,6 +3576,29 @@ declare_clippy_lint! {
+    /// let x = false;
+    /// x.then_some(0);
+    /// ```
+    #[clippy::version = "1.96.0"]
````
```suggestion
    #[clippy::version = "1.97.0"]
```

ada4a reviewed (approved):
One last change, then let's merge

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15745/7f0774a151df7239fdbf1fded3dc12d610a32e66..af4a4f953175ddb65dba3884355620bbdc43d18e)*

ada4a reviewed (approved):
Thank you:)

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/15745/da29c3e2590764f9b6e42d088bf725969b41261c..436aad4e0fa745a686c33852520e9799850e68e7)*
